src/main/java/io/rcktapp/api/handler/firehose/DeliveryStreamNameSpliterator.java

```diff
@@ -0,0 +1,53 @@
+package io.rcktapp.api.handler.firehose;
+
+import com.amazonaws.services.kinesisfirehose.AmazonKinesisFirehose;
+import com.amazonaws.services.kinesisfirehose.model.ListDeliveryStreamsRequest;
+
+import java.util.List;
+import java.util.Spliterator;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class DeliveryStreamNameSpliterator implements Spliterator<String> {
+    private final AmazonKinesisFirehose firehoseClient;
+    private String last = null;
+
+    private DeliveryStreamNameSpliterator(AmazonKinesisFirehose firehoseClient) {
+        this.firehoseClient = firehoseClient;
+    }
+
+    static Stream<String> stream(AmazonKinesisFirehose firehoseClient) {
+        return StreamSupport.stream(new DeliveryStreamNameSpliterator(firehoseClient), false);
+    }
+
+
+    @Override
+    public boolean tryAdvance(Consumer<? super String> action) {
+        List<String> page = firehoseClient.listDeliveryStreams(getRequest(last)).getDeliveryStreamNames();
+        last = page.stream().reduce((first, second) -> second).orElse(null);
+        page.forEach(action);
+        return !page.isEmpty();
+    }
```

Comment:
This is the fundamental implementation in a Spliterator - when you need to try to advance to the next element because there aren't any more, what do you do?  They pass in a consumer to which we pass the next entry(ies), and we return if this is the end of not.

and what WE do is grab a page, feed all its names it into the consumer, and remember the last entry in our spliterator instance. We return the inversion of whether the page was empty - empty - don't ask again - has content - ask again.